src/transformStateWithClones.js

```diff
@@ -7,7 +7,41 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const updatedState = [];
+  let tempObject = { ...state };
+
+  for (const action of actions) {
+    switch (true) {
+      case action.type === 'addProperties':
+        add(action.extraData);
+        break;
+      case action.type === 'removeProperties':
+        remove(action.keysToRemove);
+        break;
+      case action.type === 'clear':
+        clear();
+        break;
+    }
+  }
+
+  function add(data) {
+    tempObject = { ...tempObject, ...data };
+    updatedState.push({ ...tempObject });
```

Comment:
It's redundant to spread 'tempObject' twice, once when you're creating a new object in 'add', and again when pushing to 'updatedState'. You can simply push the result of the first spread operation to 'updatedState'.